source/h2_connection.c

```diff
@@ -845,10 +846,58 @@ int s_decoder_on_push_promise(uint32_t stream_id, uint32_t promised_stream_id, v
     return AWS_OP_SUCCESS;
 }
 
-int s_decoder_on_data(uint32_t stream_id, struct aws_byte_cursor data, void *userdata) {
+int s_decoder_on_data_begin(uint32_t stream_id, uint32_t payload_len, void *userdata) {
     struct aws_h2_connection *connection = userdata;
 
-    /* #TODO Send a window update frame back. */
+    /* A receiver that receives a flow-controlled frame MUST always account for its contribution against the connection
+     * flow-control window, unless the receiver treats this as a connection error */
+    if (connection->thread_data.window_size_self < payload_len) {
+        /* TODO: Figure out what kind error this is, I failed to find it from spec, and it is possible to happen */
+        return aws_raise_error(AWS_ERROR_INVALID_STATE);
+    }
+    connection->thread_data.window_size_self -= payload_len;
+    struct aws_h2_stream *stream;
+    if (s_get_active_stream_for_incoming_frame(connection, stream_id, AWS_H2_FRAME_T_DATA, &stream)) {
+        return AWS_OP_ERR;
+    }
+
+    if (stream) {
+        if (aws_h2_stream_on_decoder_data_begin(stream, payload_len)) {
+            return AWS_OP_ERR;
+        }
+        /* send a stream window_update frame to automatically maintain the stream self window size */
+        struct aws_h2_frame *stream_window_update_frame =
```

Comment:
[trivial] let's send the stream's WINDOW_UPDATE from stream.c

At first I wrote "Just for the sake of maintaining the logical split between stream & connection"

But later I realized the stream might decide to send RST_STREAM and close itself from within this call, so it would be a crime to send a WINDOW_UPDATE for the stream, and also the `stream` ptr itself could be invalid at this point